README.rst
Money Parser
============

.. image:: https://badge.fury.io/py/money-parser.svg
    :target: https://badge.fury.io/py/money-parser

.. image:: https://img.shields.io/pypi/pyversions/money-parser.svg
    :target: https://pypi.python.org/pypi/money-parser

Description
-----------

Money Parser is a price and currency parsing utility.
It provides methods to extract price and currency information from the raw string.
There is a lot of different price and currency formats that present values with separators, spacing etc.
This library may help you to parse such data.

Limitation
----------

 - currency with 3 numbers after decimal point (e.g. KWD, see `ISO 4217`_)

Introduction
------------

Extracting price from raw ``str``::

   >>> from money_parser import price_str
   >>> price_str('₹1,50,087.99 \n')
   '150087.99'
   >>> price_dec('+12.007')
   Decimal('12007')

Installation
------------

::

   $ pip install money-parser

The library requires Python >= 3.4

Documentation
------------------

**price_str** (raw_price, [default=<not_defined>], [dec_point='.'])

 | Search and clean price value.
 |
 | Convert raw price string presented in any localization
 | as a valid number string with an optional decimal point.
 |
 | If raw price does not contain valid price value or contains
 | more than one price value, then return default value.
 | If default value not set, then raise ValueError.
 |
 | *:param* str **raw_price:** string that contains price value.
 | *:param* **default:** value that will be returned if raw price not valid.
 | *:param* str **dec_point:** symbol that separate integer and fractional parts.
 | *:return:* cleaned price string or default value.
 | *:raise* ValueError: error if raw price not valid and default value not set.

::

    >>> price_str('+12.007')
    '12007'
    >>> price_str('- 520,05')
    '-520.05'
    >>> price_str('1,000,777.5')
    '1000777.5'
    >>> price_str('1.777.000,99')
    '1777000.99'
    >>> price_str('99,77,11,000.1')
    '997711000.1'
    >>> price_str('USD 5,242 \t\n')
    '5242'
    >>> price_str('90 210.42', dec_point='|')
    '90210|42'
    >>> price_str(None, default='0')
    '0'
    >>> price_str(42.333, default=None) is None
    True
    >>> price_str(None)
    Traceback (most recent call last):
      ...
    ValueError: Wrong raw price type "<class 'NoneType'>" (expected type "str")
    >>> price_str('')
    Traceback (most recent call last):
      ...
    ValueError: Raw price value "" does not contain valid price digits
    >>> price_str('1..2')
    Traceback (most recent call last):
      ...
    ValueError: Raw price value "1..2" contains more than one price value


**price_dec** (raw_price, [default=<not_defined>])

 | Price decimal value from raw string.
 |
 | Extract price value from input raw string and
 | present as Decimal number. Uses a price_str function for price parsing.

 | If raw price does not contain valid price value or contains
 | more than one price value, then return default value.
 | If default value not set, then raise ValueError.
 |
 | *:param* str **raw_price:** string that contains price value.
 | *:param* **default:** value that will be returned if raw price not valid.
 | *:return:* Decimal price value.
 | *:raise* ValueError: error if raw price not valid and default value not set.

::

    >>> price_dec('+12.007')
    Decimal('12007')
    >>> price_dec(': -10.99$')
    Decimal('-10.99')
    >>> price_dec('', default=Decimal('0'))
    Decimal('0')
    >>> price_dec('1..10', default=0)
    0
    >>> price_dec('410.5 - 555', default=None) is None
    True
    >>> price_dec(42.3)
    Traceback (most recent call last):
      ...
    ValueError: Wrong raw price type "<class 'float'>" (expected type "str")
    >>> price_dec('free')
    Traceback (most recent call last):
      ...
    ValueError: Raw price value "free" does not contain valid price digits
    >>> price_dec('2+2')
    Traceback (most recent call last):
      ...
    ValueError: Raw price value "2+2" contains more than one price value

Run Tests
---------

Project has tests::

    $ make test

Also available tests with coverage::

    $ make cov

Source code
-----------

The project is hosted on GitHub_

Authors and License
-------------------

The ``money-parser`` package is written by Serhii Kostel.

It's *Apache 2* licensed and freely available.


.. _`ISO 4217`: https://en.wikipedia.org/wiki/ISO_4217
.. _GitHub: https://github.com/kserhii/money-parser


CHANGES.rst
CHANGES
=======

0.0.1 (2017-12-17)
------------------

* The first release.

tests/test_price.py
from decimal import Decimal

import pytest

from money_parser import price_str, price_dec


price_str_test_cases = [
    # Integer
    ('0', '0'),
    ('42', '42'),
    ('1234', '1234'),
    ('1 234', '1234'),
    ('1 234 567', '1234567'),
    ('12.007', '12007'),
    ('12,007', '12007'),
    ('1.234.567', '1234567'),
    ('1,234,567', '1234567'),
    ('1,034.550', '1034550'),

    # Start with zero
    ('042', '42'),
    ('0.01', '0.01'),
    ('00.1', '0.1'),
    ('0 234', '234'),
    ('000 567', '567'),
    ('012.33', '12.33'),
    ('00012,33', '12.33'),

    # Signed
    ('+1', '1'),
    ('-2.99', '-2.99'),
    ('-520', '-520'),
    ('+520', '520'),
    ('- 520.05', '-520.05'),
    ('+ 520.05', '520.05'),

    # Truncated decimal
    ('99.', '99'),
    ('99,', '99'),
    ('1,099.', '1099'),
    ('1.099,', '1099'),

    # One decimal digit
    ('42.0',    '42.0'),
    ('42,0',    '42.0'),
    ('10 380,5', '10380.5'),
    ('10 380.5', '10380.5'),
    ('10.380,5', '10380.5'),
    ('10,380.5', '10380.5'),
    ('1 000 777.5', '1000777.5'),
    ('1 000 777,5', '1000777.5'),
    ('1.000.777,5', '1000777.5'),
    ('1,000,777.5', '1000777.5'),

    # Two decimal digits
    ('10.99', '10.99'),
    ('10,99', '10.99'),
    ('1234567.89', '1234567.89'),
    ('1234567,89', '1234567.89'),
    ('12 345,09', '12345.09'),
    ('12 345.09', '12345.09'),
    ('12.345,09', '12345.09'),
    ('12,345.09', '12345.09'),
    ('1 234 567.89', '1234567.89'),
    ('1 234 567,89', '1234567.89'),
    ('1.234.567,89', '1234567.89'),
    ('1,234,567.89', '1234567.89'),

    # Indian numbering system
    ('5,00,111', '500111'),
    ('5.00.111', '500111'),
    ('12,13,14,007', '121314007'),
    ('12.13.14.007', '121314007'),
    ('7,01,02,03,999', '7010203999'),
    ('7.01.02.03.999', '7010203999'),

    ('1,00,222.1', '100222.1'),
    ('1.00.222,1', '100222.1'),
    ('99,77,11,000.1', '997711000.1'),
    ('99.77.11.000,1', '997711000.1'),

    ('5,00,111.99', '500111.99'),
    ('5.00.111,99', '500111.99'),
    ('12,13,14,007.99', '121314007.99'),
    ('12.13.14.007,99', '121314007.99'),

    # Price with rubbish
    ('  42  \t \n', '42'),
    ('|620|\" ]', '620'),
    ('|620.0|\" ]', '620.0'),
    ('|620,99\'|\" ]', '620.99'),
    ('[ * |"620.99"| * ]', '620.99'),
    ('\u200e \\500 "\n \n', '500'),
    ('"855.000|', '855000'),
    ('NIO5,242', '5242'),
    ('NIO5.242', '5242'),
    ('NIO5,242.7', '5242.7'),
    ('CFA10,615', '10615'),
    ('$10.99', '10.99'),
    ('USD 10,99', '10.99'),
    ('10.99$', '10.99'),
    ('10,99USD', '10.99'),
    ('99 ₣', '99'),
    ('Price:-9.99$', '-9.99'),
    ('Price:- 9,99 ₪', '-9.99'),
    ('Start with DOT.45', '45'),
    ('Start with COMMA,45', '45'),
    ('Not a MINUS-.45', '45'),
    ('Not a PLUS+,45', '45'),
]

wrong_price_str_test_cases = [
    # Wrong type
    None, object(), b'22', 23, 24.3, Decimal('25'),
    # No a price
    '', '1...2', '+', '-', 'USD', '-NuN', '+inf',
    # Two prices
    '90 BH 210', '90-60', '451 °F = 233 °C', '50.431782|30.516382'
]


@pytest.mark.parametrize("raw_price,price", price_str_test_cases)
def test_price_str_value(raw_price, price):
    assert price == price_str(raw_price)


@pytest.mark.parametrize("wrong_raw_price", wrong_price_str_test_cases)
def test_price_str_default(wrong_raw_price):
    default = object()
    assert default == price_str(wrong_raw_price, default=default)


@pytest.mark.parametrize("wrong_raw_price", wrong_price_str_test_cases)
def test_price_str_error(wrong_raw_price):
    with pytest.raises(ValueError):
        price_str(wrong_raw_price)


def test_price_str_dec_point():
    assert '9.99' == price_str('9,99')
    assert '9|99' == price_str('9,99', dec_point='|')


def test_price_dec_value():
    assert Decimal('1') == price_dec('+1')
    assert Decimal('-10.99') == price_dec(': -10.99$')


def test_price_dec_default():
    assert Decimal('0') == price_dec('', default=Decimal('0'))
    assert 0 == price_dec('1..10', default=0)
    assert price_dec('410.5 - 555', default=None) is None


def test_price_dec_error():
    with pytest.raises(ValueError):
        price_dec('')
    with pytest.raises(ValueError):
        price_dec('7 | 128')


money_parser/__init__.py
import re
import decimal

__version__ = '0.0.1'

__all__ = ('price_str', 'price_dec',)


_CLEANED_PRICE_RE = re.compile('[+-]?(?:\d{1,3}[.,]?)+')
_FRACTIONAL_PRICE_RE = re.compile('^([\d.,]+)[.,](\d{1,2})$')

_not_defined = object()


def price_str(raw_price, default=_not_defined, dec_point='.'):
    """Search and clean price value.

    Convert raw price string presented in any localization
    as a valid number string with an optional decimal point.

    If raw price does not contain valid price value or contains
    more than one price value, then return default value.
    If default value not set, then raise ValueError.

    Examples:
        12.007          => 12007
        00012,33        => 12.33
        +1              => 1
        - 520.05        => -520.05
        1,000,777.5     => 1000777.5
        1.777.000,99    => 1777000.99
        1 234 567.89    => 1234567.89
        99.77.11.000,1  => 997711000.1
        NIO5,242        => 5242
        Not a MINUS-.45 => 45
          42  \t \n     => 42
                        => <default>
        1...2           => <default>

    :param str raw_price: string that contains price value.
    :param default: value that will be returned if raw price not valid.
    :param str dec_point: symbol that separate integer and fractional parts.
    :return: cleaned price string.
    :raise ValueError: error if raw price not valid and default value not set.
    """
    def _error_or_default(err_msg):
        if default == _not_defined:
            raise ValueError(err_msg)
        return default

    # check and clean
    if not isinstance(raw_price, str):
        return _error_or_default(
            'Wrong raw price type "{price_type}" '
            '(expected type "str")'.format(price_type=type(raw_price)))

    price = re.sub('\s', '', raw_price)
    cleaned_price = _CLEANED_PRICE_RE.findall(price)

    if len(cleaned_price) == 0:
        return _error_or_default(
            'Raw price value "{price}" does not contain '
            'valid price digits'.format(price=raw_price))

    if len(cleaned_price) > 1:
        return _error_or_default(
            'Raw price value "{price}" contains '
            'more than one price value'.format(price=raw_price))

    price = cleaned_price[0]

    # clean truncated decimal (e.g. 99. -> 99)
    price = price.rstrip('.,')

    # get sign
    sign = ''
    if price[0] in {'-', '+'}:
        sign, price = price[0], price[1:]
        sign = '-' if sign == '-' else ''

    # extract fractional digits
    fractional = _FRACTIONAL_PRICE_RE.match(price)
    if fractional:
        integer, fraction = fractional.groups()
    else:
        integer, fraction = price, ''

    # leave only digits in the integer part of the price
    integer = re.sub('\D', '', integer)

    # remove leading zeros (e.g. 007 -> 7, but 0.1 -> 0.1)
    integer = integer.lstrip('0')
    if integer == '':
        integer = '0'

    # construct price
    price = sign + integer
    if fraction:
        price = ''.join((price, dec_point, fraction))

    return price


def price_dec(raw_price, default=_not_defined):
    """Price decimal value from raw string.

    Extract price value from input raw string and
    present as Decimal number.

    If raw price does not contain valid price value or contains
    more than one price value, then return default value.
    If default value not set, then raise ValueError.

    :param str raw_price: string that contains price value.
    :param default: value that will be returned if raw price not valid.
    :return: Decimal price value.
    :raise ValueError: error if raw price not valid and default value not set.
    """
    try:
        price = price_str(raw_price)
        return decimal.Decimal(price)

    except ValueError as err:
        if default == _not_defined:
            raise err

    return default
